### ISSUE TITLE
Type of For's children function

### ISSUE THREAD

**Issue by axtgr** [LegendApp/legend-state#225]
I use For with the children function and have a couple of issues with its types:

1. The _inner_ type of the `item$` observable given to the function is converted to be optional. This is easier demonstrated with code:

```ts
const arr = observable([{ value: 1 }]) // ObservableArray<{ value: number; }[]>

return (
  <For each={arr}>
    {(item$) => { // item$ is ObservablePrimitiveBaseFns<{ value: number; } | undefined>
      return <div>{item$.value.get()}</div> // TS error due to undefined not having the value property
    }}
  </For>
) 
```

This makes working with items very unwieldy.

2. The expected return type of the children function is `ReactElement`, which permits only actual element objects and nothing else that react can render (string, number, null, etc.). I think this should've been typed as `ReactNode`.

If I ignore the errors, everything seems to be working fine and I haven't encountered a case where these restrictions would be useful (even in the optimized mode).

**Comment by jmeistrich** [comment]
1. We're working on a rewrite of the types thanks to @bram209 which will fix that. It's a problem with the ObservableArray type. But the type rewrite is tied together with a big project so it may be a few weeks until we release it.
2. Hmm, you may be right that it should be `ReactNode`. It's basically typing the children like `React.FunctionComponent` which returns a `ReactElement`. Maybe we should change that. So it's working ok to return a string but just the types are wrong?

**Comment by axtgr** [comment]
1. Good to hear that!
2. `React.FunctionComponent` actually [returns](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/657a7864cd96fc3159f8c47ac6a1d4d5330a9742/types/react/index.d.ts#L567C29-L567C29) `ReactNode` now. AFAIK they updated the type in v18 or an earlier version. Returning a string or null works fine in my ad-hoc testing, but I cannot attest that it doesn't break in some edge case.

**Comment by jmeistrich** [comment]
I think this should be fixed in the new 3.0 alpha version. Can you try updating to the latest and see if this is fixed for you?
